docs/token-contracts/fa12/4-fa12-archetype.md

```diff
@@ -0,0 +1,447 @@
+
+Note that in that case, it may be set back to 0 by having `spender` call the `transfer` entry to transfer a number of tokens equal to the remaining allowed amount, from the approver address (ie `caller` above) to the approver address (ie to itself).
+
+Indeed, according to properties [1.3](#13-no-effect-on-ledger) and [2.1](#21-effect-on-fromcaller-allowance-record) of the `transfer` entry, this has no effect on `ledger` and sets allowance record to 0.
+
+#### 4/ Effect on `operations`
+```archetype
+length (operations) = 0
+```
+> No operation generated.
+
+### getAllowance
+
+```archetype
+getter getAllowance (owner : address, spender : address) : nat {
+  return (allowance[(owner, spender)].amount)
+}
+```
+
+#### 1/ Effect on `ledger`
+```archetype
+ledger = before.ledger
+```
+> No effect on `ledger`.
+
+#### 2/ Effect on `allowance`
+```archetype
+allowance = before.allowance
+```
+> No effect on `allowance`.
+
+#### 3/ Explicit `fail`
+No explicit fail. The entry implicitely fails though if the provided callback is invalid.
+
+#### 4/ Effect on `operations`
+```archetype
+length (operations) = 1
+```
+> Creates one callback operation.
+
+### getBalance
+
+```archetype
+getter getBalance (owner : address) : nat {
+  return (ledger[owner].tokens)
+}
+```
+
+#### 1/ Effect on `ledger`
+```archetype
+ledger = before.ledger
+```
+> No effect on `ledger`.
+
+#### 2/ Effect on `allowance`
+```archetype
+allowance = before.allowance
+```
+> No effect on `allowance`.
+
+#### 3/ Explicit fail
+No explicit fail. The entry implicitely fails though if the provided callback is invalid.
+
+#### 4/ Effect on `operations`
+```archetype
+length (operations) = 1
+```
+> Creates one callback operation.
+
+
+### getTotalSupply
+
+```archetype
+getter getTotalSupply () : nat {
+  return totalsupply
+}
+```
+
+#### 1/ Effect on `ledger`
+```archetype
+ledger = before.ledger
+```
+> No effect on `ledger`.
+
+#### 2/ Effect on `allowance`
+```archetype
+allowance = before.allowance
+```
+> No effect on `allowance`.
+
+#### 3/ Explicit `fail`
+No explicit fail. The entry implicitely fails though if the provided callback is invalid.
+
+#### 4/ Effect on `operations`
+```archetype
+length (operations) = 1
+```
+> Creates one callback operation.
+
+
+## Verify properties
+This section presents how to verify the contract properties with Why3.
+
+The following command generates the whyml version of the contract:
+```
+$ archetype -t whyml fa12.arl > fa12.mlw
+```
+
+The following command launches the why3 IDE on the whyml version:
+```
+$ why3 ide -L $OPAM_SWITCH_PREFIX/share/archetype/mlw fa12.mlw
+```
+
+It may also be launched from VS Code with the following command:
+```
+Archetype: Verify with Why3
+```
+> Install [why3](https://docs.archetype-lang.org/getting-started-1#verification-tools)
+
+Right-click on the `Fa12` module in the left-hand panel, and select `Auto level 2` as the solving strategy:
```

Comment:
It'd be nice to put a couple of sentences summarizing what is happening, like: "The annotated WhyML program representing the contract is used to generate proof-goal scripts for ... and then they are run with timeouts ..." etc

Reply:
Done.